File: packages/srlc/test/golden/customer_protection.kt
package com.summit.srlc

import java.security.MessageDigest
import org.apache.kafka.streams.kstream.KStream

object SrlcHelpers {
  private fun assertFormat(value: String?, format: String) {
    if (value == null) return
    when (format) {
      "ssn" -> require(Regex("^[0-9]{3}-?[0-9]{2}-?[0-9]{4}$").matches(value)) { "SRLC format violation for SSN: $value" }
      "iban" -> require(Regex("^[A-Z0-9]{15,34}$").matches(value)) { "SRLC format violation for IBAN: $value" }
      "phone" -> require(Regex("^\+?[0-9]{10,15}$").matches(value)) { "SRLC format violation for phone: $value" }
    }
  }

  fun mask(value: String?, keep: Int, maskChar: Char, format: String): String? {
    if (value == null) return null
    assertFormat(value, format)
    if (keep <= 0) {
      return value.replace(Regex("[A-Za-z0-9]"), maskChar.toString())
    }
    val builder = StringBuilder()
    var visible = 0
    for (ch in value.reversed()) {
      if (ch.isLetterOrDigit()) {
        if (visible < keep) {
          builder.insert(0, ch)
          visible += 1
        } else {
          builder.insert(0, maskChar)
        }
      } else {
        builder.insert(0, ch)
      }
    }
    return builder.toString()
  }

  fun hash(value: String?, algorithm: String, saltScope: String): String? {
    if (value == null) return null
    val salt = if (saltScope == "global") "SRLC_GLOBAL" else "SRLC_SESSION"
    val digest = MessageDigest.getInstance(algorithm.uppercase())
    val hashed = digest.digest((value + salt).toByteArray())
    return hashed.joinToString("") { byte -> "%02x".format(byte) }
  }

  fun tokenize(value: String?, namespace: String, preserveFormat: Boolean, format: String): String? {
    if (value == null) return null
    assertFormat(value, format)
    val digest = MessageDigest.getInstance("SHA-256")
    val hashed = digest.digest(("$namespace:" + value).toByteArray())
    val token = hashed.joinToString("") { byte -> "%02x".format(byte) }
    return if (preserveFormat) token.substring(0, value.length) else token
  }

  fun generalize(value: String?, granularity: String): String? {
    if (value == null) return null
    if (granularity == "none") return value
    return "$granularity::" + value
  }
}

fun KStream<String, MutableMap<String, Any?>>.applySrlcCustomerProtection(): KStream<String, MutableMap<String, Any?>> {
  return this.mapValues { value ->
    val result = value.toMutableMap()
    var current0 = value["customer.ssn"] as? String
    current0 = SrlcHelpers.mask(current0, 4, '#', "ssn")
    current0 = SrlcHelpers.hash(current0, "sha256", "session")
    result["customer.ssn"] = current0
    // customer.ssn:mask(format=ssn,keep=4,char=#) |> hash(format=ssn,algorithm=sha256,salt=session)
    var current1 = value["account.iban"] as? String
    current1 = SrlcHelpers.tokenize(current1, "payments", true, "iban")
    result["account.iban"] = current1
    // account.iban:tokenize(format=iban,namespace=payments,preserveFormat=true)
    var current2 = value["contact.phone"] as? String
    current2 = SrlcHelpers.mask(current2, 4, '#', "phone")
    current2 = SrlcHelpers.generalize(current2, "region")
    result["contact.phone"] = current2
    // contact.phone:mask(format=phone,keep=4,char=#) |> generalize(format=phone,granularity=region)
    result
  }
}

File: packages/srlc/test/golden/customer_protection.ts
import { createHash } from 'node:crypto';

export type SrlcRecord = Record<string, unknown>;

type NullableString = string | null;

function assertFormat(value: NullableString, format: string): void {
  if (value == null) {
    return;
  }
  const patterns: Record<string, RegExp> = {
    ssn: /^[0-9]{3}-?[0-9]{2}-?[0-9]{4}$/,
    iban: /^[A-Z0-9]{15,34}$/,
    phone: /^\+?[0-9]{10,15}$/,
  };
  const pattern = patterns[format];
  if (pattern && !pattern.test(value)) {
    throw new Error(`SRLC format violation for ${format}: ${value}`);
  }
}

function mask(value: NullableString, keep: number, maskChar: string, format: string): NullableString {
  if (value == null) {
    return value;
  }
  assertFormat(value, format);
  if (keep <= 0) {
    return value.replace(/[A-Za-z0-9]/g, maskChar);
  }
  let visible = 0;
  let result = '';
  for (let idx = value.length - 1; idx >= 0; idx -= 1) {
    const ch = value[idx];
    if (/[A-Za-z0-9]/.test(ch)) {
      if (visible < keep) {
        result = ch + result;
        visible += 1;
      } else {
        result = maskChar + result;
      }
    } else {
      result = ch + result;
    }
  }
  return result;
}

function hash(value: NullableString, algorithm: 'sha256' | 'sha512', saltScope: string): NullableString {
  if (value == null) {
    return value;
  }
  const salt = saltScope === 'global' ? 'SRLC_GLOBAL' : 'SRLC_SESSION';
  return createHash(algorithm).update(value + salt).digest('hex');
}

function tokenize(value: NullableString, namespace: string, preserveFormat: boolean, format: string): NullableString {
  if (value == null) {
    return value;
  }
  assertFormat(value, format);
  const token = createHash('sha256').update(`${namespace}:${value}`).digest('hex');
  return preserveFormat ? token.slice(0, value.length) : token;
}

function generalize(value: NullableString, granularity: string): NullableString {
  if (value == null) {
    return value;
  }
  if (granularity === 'none') {
    return value;
  }
  return `${granularity}::${value}`;
}

export const helpers = {
  assertFormat,
  mask,
  hash,
  tokenize,
  generalize
};

export function applyCustomerProtectionRedactions(record: SrlcRecord): SrlcRecord {
  const next: SrlcRecord = { ...record };

  {
    const original = record['customer.ssn'];
    let current0: NullableString = original == null ? null : String(original);
    current0 = mask(current0, 4, '#', 'ssn');
    current0 = hash(current0, 'sha256', 'session');
    next['customer.ssn'] = current0;
    // customer.ssn:mask(format=ssn,keep=4,char=#) |> hash(format=ssn,algorithm=sha256,salt=session)
  }

  {
    const original = record['account.iban'];
    let current1: NullableString = original == null ? null : String(original);
    current1 = tokenize(current1, 'payments', true, 'iban');
    next['account.iban'] = current1;
    // account.iban:tokenize(format=iban,namespace=payments,preserveFormat=true)
  }

  {
    const original = record['contact.phone'];
    let current2: NullableString = original == null ? null : String(original);
    current2 = mask(current2, 4, '#', 'phone');
    current2 = generalize(current2, 'region');
    next['contact.phone'] = current2;
    // contact.phone:mask(format=phone,keep=4,char=#) |> generalize(format=phone,granularity=region)
  }

  return next;
}